File: integrations/strands-ts/src/policy.ts
/**
 * policy.ts — Native-import policy gate for BeforeToolCallEvent.
 *
 * Primary binding: NATIVE import of scripts/hooks/config-protection.js via
 * its module.exports.run(raw, opts) API — no subprocess.  This is the key
 * differentiator from the Python adapter which uses a subprocess binding.
 *
 * The run() API contract (from run-hook.js §8.2, Form 2):
 *   run(rawJsonString, { truncated, maxStdin }) → string | { exitCode, stderr?, stdout? }
 *
 * Exit code / return value semantics:
 *   exitCode 0   → allow (return null)
 *   exitCode 2   → block (return block reason string from stderr/stdout)
 *   other/error  → fail-open (return null)
 *
 * Engine location resolution (in priority order):
 *   1. Constructor option `engineRoot` (explicit root dir — if provided but invalid,
 *      stops here and returns null; does NOT fall through to auto-discovery).
 *   2. FLOW_AGENTS_ENGINE_ROOT env var.
 *   3. Relative to this file: ../../../scripts/hooks/ (repo checkout layout).
 *   4. Walk up from cwd: node_modules/@kontourai/flow-agents/scripts/hooks/.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ---------------------------------------------------------------------------
// Protected files set — mirrors PROTECTED_FILES in config-protection.js
// Used as fallback when native engine is unavailable.
// ---------------------------------------------------------------------------

export const PROTECTED_FILES: ReadonlySet<string> = new Set([
  // ESLint
  ".eslintrc",
  ".eslintrc.js",
  ".eslintrc.cjs",
  ".eslintrc.json",
  ".eslintrc.yml",
  ".eslintrc.yaml",
  "eslint.config.js",
  "eslint.config.mjs",
  "eslint.config.cjs",
  "eslint.config.ts",
  "eslint.config.mts",
  "eslint.config.cts",
  // Prettier
  ".prettierrc",
  ".prettierrc.js",
  ".prettierrc.cjs",
  ".prettierrc.json",
  ".prettierrc.yml",
  ".prettierrc.yaml",
  "prettier.config.js",
  "prettier.config.cjs",
  "prettier.config.mjs",
  // Biome
  "biome.json",
  "biome.jsonc",
  // Ruff
  ".ruff.toml",
  "ruff.toml",
  // Others
  ".shellcheckrc",
  ".stylelintrc",
  ".stylelintrc.json",
  ".stylelintrc.yml",
  ".markdownlint.json",
  ".markdownlint.yaml",
  ".markdownlintrc",
]);

// Write-like tool names — only gate write-like tools; reads are always allowed.
const WRITE_TOOLS: ReadonlySet<string> = new Set([
  "edit",
  "write",
  "fs_write",
  "apply_patch",
  "create_file",
  "str_replace_editor",
]);

const BLOCK_REASON_TEMPLATE = (basename: string): string =>
  `BLOCKED: Modifying ${basename} is not allowed. ` +
  "Fix the source code to satisfy linter/formatter rules instead of " +
  "weakening the config. If this is a legitimate config change, " +
  "disable the config-protection policy gate temporarily.";

// ---------------------------------------------------------------------------
// Engine location resolver
// ---------------------------------------------------------------------------

/**
 * Resolve the hooksDir (directory containing config-protection.js etc).
 *
 * When engineRootOverride is provided:
 *   - If the override is a valid repo root (contains scripts/hooks/run-hook.js):
 *     return that hooksDir.
 *   - If the override is a valid hooks dir itself (contains run-hook.js directly):
 *     return that dir.
 *   - Otherwise: return null. Do NOT fall through to auto-discovery.
 *     This ensures an explicit but invalid override surfaces clearly.
 *
 * When engineRootOverride is absent:
 *   Auto-discover via env var → file-relative → cwd walk.
 */
function resolveHooksDir(engineRootOverride?: string): string | null {
  // 1. Explicit override provided — do not fall through to auto-discovery
  if (engineRootOverride !== undefined) {
    const asRepoRoot = path.join(engineRootOverride, "scripts", "hooks");
    if (fs.existsSync(path.join(asRepoRoot, "run-hook.js"))) {
      return asRepoRoot;
    }
    // Allow engineRoot to point directly at scripts/hooks/
    if (fs.existsSync(path.join(engineRootOverride, "run-hook.js"))) {
      return engineRootOverride;
    }
    return null; // explicit override specified but invalid
  }

  // 2. Env var
  const envRoot = process.env.FLOW_AGENTS_ENGINE_ROOT;
  if (envRoot) {
    const candidate = path.join(envRoot, "scripts", "hooks");
    if (fs.existsSync(path.join(candidate, "run-hook.js"))) return candidate;
    if (fs.existsSync(path.join(envRoot, "run-hook.js"))) return envRoot;
  }

  // 3. Relative to this file (repo checkout layout)
  // __dirname is dist/src/ at runtime, src/ when run as source
  // scripts/hooks/ is relative to the repo root, not this file
  for (const relPath of [
    path.resolve(__dirname, "../../../scripts/hooks"),     // src/ layout (3 up)
    path.resolve(__dirname, "../../../../scripts/hooks"),  // dist/src/ layout (4 up)
  ]) {
    if (fs.existsSync(path.join(relPath, "run-hook.js"))) {
      return relPath;
    }
  }

  // 4. Walk up from cwd looking for npm-installed package
  let current = process.cwd();
  for (let i = 0; i < 10; i++) {
    const candidate = path.join(
      current,
      "node_modules",
      "@kontourai",
      "flow-agents",
      "scripts",
      "hooks"
    );
    if (fs.existsSync(path.join(candidate, "run-hook.js"))) {
      return candidate;
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }

  return null;
}

// ---------------------------------------------------------------------------
// Native engine call (Form 2 — module.exports.run())
// ---------------------------------------------------------------------------

interface HookRunOutput {
  exitCode?: number;
  stderr?: string;
  stdout?: string;
}

type HookRunResult = string | HookRunOutput;

interface HookModule {
  run(raw: string, options: { truncated: boolean; maxStdin: number }): HookRunResult;
}

function loadHookModule(hooksDir: string, scriptName: string): HookModule | null {
  const scriptPath = path.join(hooksDir, scriptName);
  if (!fs.existsSync(scriptPath)) return null;

  try {
    // Use createRequire for CJS modules from ESM context
    const require = createRequire(import.meta.url);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const mod = require(scriptPath) as any;
    if (mod && typeof mod.run === "function") {
      return mod as HookModule;
    }
    return null;
  } catch {
    return null;
  }
}

function interpretRunResult(result: HookRunResult): { exitCode: number; blockReason: string | null } {
  if (typeof result === "string") {
    // String return = stdout pass-through → allow
    return { exitCode: 0, blockReason: null };
  }
  if (result && typeof result === "object") {
    const exitCode = typeof result.exitCode === "number" ? result.exitCode : 0;
    if (exitCode === 2) {
      const reason =
        result.stderr?.trim() ||
        result.stdout?.trim() ||
        "BLOCKED: config-protection policy blocked this action.";
      return { exitCode: 2, blockReason: reason };
    }
    return { exitCode, blockReason: null };
  }
  return { exitCode: 0, blockReason: null };
}

// ---------------------------------------------------------------------------
// Pure-TS fallback — mirrors the JS config-protection.js logic
// ---------------------------------------------------------------------------

function tsConfigProtection(
  toolName: string,
  toolInput: Record<string, unknown>,
  protectedFiles: ReadonlySet<string> = PROTECTED_FILES
): string | null {
  if (!WRITE_TOOLS.has(toolName.toLowerCase())) return null;

  const filePath =
    (toolInput.path as string | undefined) ||
    (toolInput.file_path as string | undefined) ||
    "";
  if (!filePath) return null;

  const basename = path.basename(filePath);
  if (protectedFiles.has(basename)) {
    return BLOCK_REASON_TEMPLATE(basename);
  }
  return null;
}

// ---------------------------------------------------------------------------
// PolicyGate
// ---------------------------------------------------------------------------

export interface PolicyGateOptions {
  /**
   * Root directory of the @kontourai/flow-agents package (the directory
   * containing scripts/hooks/run-hook.js). Defaults to auto-discovery.
   *
   * If provided but invalid (engine not found there), the gate uses pure-TS
   * fallback and does NOT attempt auto-discovery.
   */
  engineRoot?: string;
  /**
   * Custom set of protected file basenames. When provided, the native engine
   * is bypassed and pure-TS evaluation is used (intended for tests only).
   */
  customProtectedFiles?: ReadonlySet<string>;
  /**
   * If true, suppress the one-time console.warn when falling back to TS evaluation.
   */
  suppressFallbackWarning?: boolean;
}

export class PolicyGate {
  private readonly configProtectionModule: HookModule | null;
  private readonly customProtectedFiles: ReadonlySet<string> | undefined;
  private readonly suppressFallbackWarning: boolean;
  private _warnedFallback = false;

  constructor(options: PolicyGateOptions = {}) {
    this.customProtectedFiles = options.customProtectedFiles;
    this.suppressFallbackWarning = options.suppressFallbackWarning ?? false;

    const hooksDir = resolveHooksDir(options.engineRoot);
    if (hooksDir) {
      this.configProtectionModule = loadHookModule(hooksDir, "config-protection.js");
    } else {
      this.configProtectionModule = null;
    }
  }

  get engineAvailable(): boolean {
    return this.configProtectionModule !== null;
  }

  private warnFallback(): void {
    if (this._warnedFallback || this.suppressFallbackWarning) return;
    this._warnedFallback = true;
    console.warn(
      "[flow-agents-strands] Warning: The Flow Agents policy engine (config-protection.js) " +
        "could not be loaded via native import. Policy gates are degrading to the built-in " +
        "TypeScript fallback (fail-open for unknown cases). " +
        "Set FLOW_AGENTS_ENGINE_ROOT to the @kontourai/flow-agents root directory to " +
        "use the canonical engine. See README.md §Limitations."
    );
  }

  /**
   * Check whether a tool call is allowed by config-protection policy.
   *
   * Returns null if allowed, or a block-reason string if blocked.
   * Always fail-open on errors.
   */
  checkToolCall(
    toolName: string,
    toolInput: Record<string, unknown>
  ): string | null {
    // Tool-name pre-filter: reads are always allowed
    if (!WRITE_TOOLS.has(toolName.toLowerCase())) return null;

    // Custom protected set → pure-TS evaluation only
    if (this.customProtectedFiles !== undefined) {
      return tsConfigProtection(toolName, toolInput, this.customProtectedFiles);
    }

    // Native engine call (the key differentiator vs Python subprocess adapter)
    if (this.configProtectionModule !== null) {
      const payload = JSON.stringify({
        hook_event_name: "PreToolUse",
        tool_name: toolName,
        tool_input: toolInput,
      });

      try {
        const result = this.configProtectionModule.run(payload, {
          truncated: false,
          maxStdin: 1024 * 1024,
        });
        const { blockReason } = interpretRunResult(result);
        return blockReason;
      } catch {
        // fail-open on native call errors
        return null;
      }
    }

    // Fallback: pure-TS evaluation
    this.warnFallback();
    return tsConfigProtection(toolName, toolInput);
  }
}

File: integrations/strands/flow_agents_strands/policy.py
"""
policy.py — Policy gates for BeforeToolCallEvent.

Primary binding: subprocess to the canonical Node.js engine
(scripts/hooks/run-hook.js → config-protection.js) for the authoritative
policy decision.  The tool-name pre-filter (write-like tools only) is applied
in Python before calling the engine, since the engine itself does not filter by
tool name — it blocks on the file basename alone.

Fallback: if Node.js is absent or the engine script cannot be located, the gate
degrades to the pure-Python implementation of the same logic and emits a
one-time RuntimeWarning.  See README.md §Limitations.

Engine contract (contract_version "1.0"):
  - Payload: JSON on stdin with hook_event_name, tool_name, tool_input fields.
  - Exit code 0 = allow, exit code 2 = block, other = error (fail-open).
  - Stderr carries the block reason when exit code is 2.
  - See docs/spec/runtime-hook-surface.md §8 for the full contract.

Custom protected_files: when the caller passes a non-default ``protected_files``
frozenset, the engine subprocess is bypassed and the Python evaluation is used
directly (the subprocess cannot receive a runtime-custom set).
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import warnings
from pathlib import Path
from typing import Optional

# ---------------------------------------------------------------------------
# Locate the canonical engine — supports both installed-package and repo layouts.
# ---------------------------------------------------------------------------

def _find_engine_paths() -> tuple[Optional[str], Optional[str]]:
    """
    Return (node_executable, run_hook_path) or (None, None) if unavailable.

    Search order for run-hook.js:
      1. Env var FLOW_AGENTS_ENGINE_PATH (explicit override).
      2. Relative to this file: ../../../../scripts/hooks/run-hook.js
         (works when running from the source repo checkout).
      3. Installed via npm: walk up from CWD looking for
         node_modules/@kontourai/flow-agents/scripts/hooks/run-hook.js
    """
    node = shutil.which("node")
    if not node:
        return None, None

    # 1. Explicit override
    env_path = os.environ.get("FLOW_AGENTS_ENGINE_PATH")
    if env_path:
        p = Path(env_path)
        if p.is_file():
            return node, str(p)

    # 2. Relative to this source file (repo checkout: integrations/strands/flow_agents_strands/)
    candidate = Path(__file__).parent.parent.parent.parent / "scripts" / "hooks" / "run-hook.js"
    if candidate.is_file():
        return node, str(candidate)

    # 3. npm-installed package layout
    cwd = Path.cwd()
    for parent in [cwd, *cwd.parents]:
        candidate = (
            parent / "node_modules" / "@kontourai" / "flow-agents" / "scripts" / "hooks" / "run-hook.js"
        )
        if candidate.is_file():
            return node, str(candidate)

    return node, None


# Resolved at module import time; callers can override via constructor parameters.
_NODE_BIN, _RUN_HOOK_PATH = _find_engine_paths()


# Sentinel for "not provided" constructor parameters
_UNSET = object()

# ---------------------------------------------------------------------------
# Pure-Python fallback — mirrors config-protection.js PROTECTED_FILES
# ---------------------------------------------------------------------------

PROTECTED_FILES = frozenset(
    [
        # ESLint
        ".eslintrc",
        ".eslintrc.js",
        ".eslintrc.cjs",
        ".eslintrc.json",
        ".eslintrc.yml",
        ".eslintrc.yaml",
        "eslint.config.js",
        "eslint.config.mjs",
        "eslint.config.cjs",
        "eslint.config.ts",
        "eslint.config.mts",
        "eslint.config.cts",
        # Prettier
        ".prettierrc",
        ".prettierrc.js",
        ".prettierrc.cjs",
        ".prettierrc.json",
        ".prettierrc.yml",
        ".prettierrc.yaml",
        "prettier.config.js",
        "prettier.config.cjs",
        "prettier.config.mjs",
        # Biome
        "biome.json",
        "biome.jsonc",
        # Ruff
        ".ruff.toml",
        "ruff.toml",
        # Others
        ".shellcheckrc",
        ".stylelintrc",
        ".stylelintrc.json",
        ".stylelintrc.yml",
        ".markdownlint.json",
        ".markdownlint.yaml",
        ".markdownlintrc",
    ]
)

_BLOCK_REASON_TEMPLATE = (
    "BLOCKED: Modifying {basename} is not allowed. "
    "Fix the source code to satisfy linter/formatter rules instead of "
    "weakening the config. If this is a legitimate config change, "
    "disable the config-protection policy gate temporarily."
)

# Write-like tool names — both the engine and the Python fallback only gate
# write-like tools.  Reads on protected files are always allowed.
_WRITE_TOOLS = frozenset(
    {
        "edit",
        "write",
        "fs_write",
        "apply_patch",
        "create_file",
        "str_replace_editor",
    }
)


def _python_config_protection(
    tool_name: str,
    tool_input: dict,
    protected: frozenset = PROTECTED_FILES,
) -> Optional[str]:
    """
    Pure-Python evaluation of config-protection.

    Mirrors the logic in scripts/hooks/config-protection.js.
    Called when Node.js is unavailable or a custom protected set is in use.
    """
    if tool_name.lower() not in _WRITE_TOOLS:
        return None
    file_path = tool_input.get("path") or tool_input.get("file_path") or ""
    if not file_path:
        return None
    basename = Path(file_path).name
    if basename in protected:
        return _BLOCK_REASON_TEMPLATE.format(basename=basename)
    return None


# ---------------------------------------------------------------------------
# Subprocess binding to the engine contract
# ---------------------------------------------------------------------------

def _invoke_engine(
    hook_id: str,
    hook_script: str,
    payload: dict,
    *,
    node_bin: str,
    run_hook_path: str,
    extra_env: Optional[dict] = None,
) -> tuple[int, str, str]:
    """
    Invoke the canonical engine via subprocess.

    Returns (exit_code, stdout, stderr).
    On subprocess errors returns (1, "", "<error>") so callers can fail-open.
    """
    env = dict(os.environ)
    if extra_env:
        env.update(extra_env)
    env["FLOW_AGENTS_HOOK_RUNTIME"] = "strands"

    # run-hook.js resolves hook_script relative to its own directory
    hooks_dir = str(Path(run_hook_path).parent)

    try:
        result = subprocess.run(
            [node_bin, run_hook_path, hook_id, hook_script],
            input=json.dumps(payload),
            capture_output=True,
            text=True,
            timeout=15,
            env=env,
            cwd=hooks_dir,
        )
        return result.returncode, result.stdout, result.stderr
    except subprocess.TimeoutExpired:
        return 1, "", "[policy] engine subprocess timed out"
    except OSError as exc:
        return 1, "", f"[policy] engine subprocess failed: {exc}"


# ---------------------------------------------------------------------------
# PolicyGate
# ---------------------------------------------------------------------------


class PolicyGate:
    """
    Evaluates tool-call policy gates.

    Designed for use inside BeforeToolCallEvent callbacks; does NOT depend on
    any Strands import so it is fully testable without the SDK installed.

    Primary mode: spawns ``node run-hook.js config-protection.js`` to delegate
    to the engine contract (contract_version "1.0").  The tool-name pre-filter
    is applied in Python before calling the engine.

    Fallback mode: if Node.js or the engine script is unavailable, degrades to
    the built-in Python implementation and emits a one-time RuntimeWarning.

    Custom protected_files: if a non-default frozenset is passed, Python
    evaluation is used directly (the engine subprocess cannot accept a
    runtime-custom set).  This is intended for tests and local override only.

    The ``_node_bin`` and ``_run_hook_path`` constructor parameters allow tests
    to inject fakes without touching the filesystem.
    """

    def __init__(
        self,
        protected_files: Optional[frozenset] = None,
        _node_bin: object = _UNSET,
        _run_hook_path: object = _UNSET,
    ) -> None:
        # Allow tests to inject explicit values (including None to force fallback).
        # _UNSET means "use the module-level resolved value".
        self._node_bin: Optional[str] = (
            _node_bin if _node_bin is not _UNSET else _NODE_BIN  # type: ignore[assignment]
        )
        self._run_hook_path: Optional[str] = (
            _run_hook_path if _run_hook_path is not _UNSET else _RUN_HOOK_PATH  # type: ignore[assignment]
        )
        # Custom protected set: bypass engine subprocess, use Python directly.
        self._custom_protected: Optional[frozenset] = protected_files
        self._warned_fallback = False

    @property
    def _engine_available(self) -> bool:
        return bool(self._node_bin and self._run_hook_path)

    def _warn_fallback(self) -> None:
        if not self._warned_fallback:
            self._warned_fallback = True
            warnings.warn(
                "flow-agents-strands: Node.js or the Flow Agents engine script is "
                "not available. Policy gates are degrading to the built-in Python "
                "fallback (fail-open for unknown cases). Install Node.js and ensure "
                "the @kontourai/flow-agents package is reachable to use the canonical "
                "engine contract. See README.md §Limitations.",
                RuntimeWarning,
                stacklevel=4,
            )

    def check_tool_call(
        self,
        tool_name: str,
        tool_input: dict,
    ) -> Optional[str]:
        """
        Evaluate policy for a tool call.

        Returns None if the call is allowed, or a block-reason string if it
        should be cancelled (the string becomes the cancel_tool message).
        """
        return self._check_config_protection(tool_name, tool_input)

    def _check_config_protection(
        self,
        tool_name: str,
        tool_input: dict,
    ) -> Optional[str]:
        """
        Invoke config-protection.

        Tool-name pre-filter: only write-like tools are gated.  Reads on
        protected files are always allowed (this is the same rule as the JS
        engine — the engine short-circuits for the same reason; the pre-filter
        avoids the subprocess round-trip for non-write tools).

        If a custom protected_files set was passed to the constructor, Python
        evaluation is used instead of the engine subprocess.

        If Node.js or the engine script is unavailable, falls back to Python
        evaluation with the default PROTECTED_FILES set.
        """
        # Tool-name pre-filter (read tools always allowed)
        if tool_name.lower() not in _WRITE_TOOLS:
            return None

        # Custom protected set → Python evaluation only
        if self._custom_protected is not None:
            return _python_config_protection(tool_name, tool_input, self._custom_protected)

        # Engine subprocess → authoritative decision
        if self._engine_available:
            payload = {
                "hook_event_name": "PreToolUse",
                "tool_name": tool_name,
                "tool_input": tool_input,
            }

            exit_code, stdout, stderr = _invoke_engine(
                hook_id="config-protection",
                hook_script="config-protection.js",
                payload=payload,
                node_bin=self._node_bin,
                run_hook_path=self._run_hook_path,
            )

            if exit_code == 2:
                reason = stderr.strip() or stdout.strip() or "BLOCKED: config-protection policy blocked this action."
                return reason

            if exit_code == 0:
                return None

            # Engine error (exit_code not 0 or 2) → fail-open
            return None

        # No engine available → Python fallback
        self._warn_fallback()
        return _python_config_protection(tool_name, tool_input)
